lib/bye_bye_bye/utils.ex

```diff
@@ -0,0 +1,173 @@
+    }
+  end
+
+  @doc """
+  Gets schedules affected by a cancellation alert.
+  Returns a map of trip_id to schedule entries.
+
+  ## Parameters
+    * `alert` - The alert to process
+    * `now` - Current time as DateTime, used to determine service day
+  """
+  def get_affected_schedules(alert, now) do
+    period_params =
+      alert["attributes"]["active_period"]
+      |> Enum.map(&period_params(&1, now))
+      |> Enum.filter(& &1)
+
+    trips =
+      alert["attributes"]["informed_entity"]
+      |> Enum.filter(fn entity -> !entity["stop"] && entity["trip"] end)
+      |> Enum.map(& &1["trip"])
+      |> then(fn trips ->
+        case trips do
+          [] -> []
+          _ -> [%{trips: trips}]
+        end
+      end)
+
+    routes =
+      alert["attributes"]["informed_entity"]
+      |> Enum.filter(fn entity -> !entity["stop"] && !entity["trip"] && entity["route"] end)
+      |> Enum.map(& &1["route"])
+      |> then(fn routes ->
+        case routes do
+          [] -> []
+          _ -> [%{routes: routes}]
+        end
+      end)
+
+    params = trips ++ routes
+
+    params
+    |> Enum.flat_map(&Enum.map(period_params, fn params -> Map.merge(params, &1) end))
+    |> Enum.map(&MbtaClient.get_schedules/1)
+    |> Enum.map(fn {:ok, schedules} -> schedules end)
+    |> Enum.reduce(%{}, &Map.merge(&1, &2))
+  end
+
+  defp service_day_times(now) do
+    now = DateTime.shift_zone!(now, "America/New_York")
+    time = DateTime.to_time(now)
+
+    if Time.before?(time, ~T[03:00:00]) do
+      {
+        now
+        |> DateTime.add(-1, :day)
+        |> DateTime.to_date()
+        |> DateTime.new!(~T[03:00:00], "America/New_York"),
+        now |> DateTime.to_date() |> DateTime.new!(~T[02:59:59], "America/New_York")
+      }
+    else
+      {
+        DateTime.to_date(now) |> DateTime.new!(~T[03:00:00], "America/New_York"),
+        DateTime.add(now, 1, :day)
+        |> DateTime.to_date()
+        |> DateTime.new!(~T[02:59:59], "America/New_York")
+      }
+    end
+  end
+
+  defp to_gtfs_time_string(datetime) do
+    time =
+      datetime
+      |> DateTime.shift_zone!("America/New_York")
+      |> DateTime.to_time()
+
+    hour = if time.hour < 3, do: time.hour + 24, else: time.hour
+
+    hour = hour |> Integer.to_string() |> String.pad_leading(2, "0")
+    minute = time.minute |> Integer.to_string() |> String.pad_leading(2, "0")
+    second = time.second |> Integer.to_string() |> String.pad_leading(2, "0")
+
+    "#{hour}:#{minute}:#{second}"
+  end
+
+  defp period_params(active_period, now) do
+    {:ok, start_time, _} = DateTime.from_iso8601(active_period["start"])
+    start_time = DateTime.shift_zone!(start_time, "America/New_York")
+
+    {:ok, end_time, _} = DateTime.from_iso8601(active_period["end"])
+    end_time = DateTime.shift_zone!(end_time, "America/New_York")
+
+    if current_service_day?(start_time, fn -> now end) or
+         current_service_day?(end_time, fn -> now end) do
```

Comment:
action needed:
Wouldn't we still want to handle the alert if the current service day is >= the start time's, and <= the end time's service day?
If I'm reading correctly it looks like this code will only keep alerts that either start or end on the current service day.

(Is there an unspoken rule that cancelation alerts never span more than one service day?)

Reply:
Nope, no rule, you're right. I think I overthought this function. The new version computes the overlap with the current service day, if there is no overlap it'll return nil, otherwise it'll use the overlap as the time period.

This should support alerts with a date range including the current service day (the resulting overlap will be the full day) and as a bonus I think is much easier to understand.